lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,652 @@
+    #[clap(long)]
+    pub token: Option<String>,
+    /// Skip validation of the uploaded package
+    #[clap(long)]
+    pub no_validate: bool,
+    /// Directory containing the `wasmer.toml` (defaults to current root dir)
+    #[clap(name = "PACKAGE_PATH")]
+    pub package_path: Option<String>,
+}
+
+#[derive(Debug, Error)]
+enum PublishError {
+    #[error("Cannot publish without a module.")]
+    NoModule,
+    #[error("Unable to publish the \"{module}\" module because \"{}\" is not a file", path.display())]
+    SourceMustBeFile { module: String, path: PathBuf },
+    #[error("Unable to load the bindings for \"{module}\" because \"{}\" doesn't exist", path.display())]
+    MissingBindings { module: String, path: PathBuf },
+    #[error("Error building package when parsing module \"{0}\": {1}.")]
+    ErrorBuildingPackage(String, io::Error),
+    #[error(
+        "Path \"{0}\", specified in the manifest as part of the package file system does not exist.",
+    )]
+    MissingManifestFsPath(String),
+    #[error("When processing the package filesystem, found path \"{0}\" which is not a directory")]
+    PackageFileSystemEntryMustBeDirectory(String),
+}
+
+impl Publish {
+    /// Executes `wasmer publish`
+    pub fn execute(&self) -> Result<(), anyhow::Error> {
+        let mut builder = Builder::new(Vec::new());
+
+        let cwd = match self.package_path.as_ref() {
+            Some(s) => std::env::current_dir()?.join(s),
+            None => std::env::current_dir()?,
+        };
+
+        let manifest_path_buf = cwd.join("wasmer.toml");
+        let manifest = std::fs::read_to_string(&manifest_path_buf)
+            .map_err(|e| anyhow::anyhow!("could not find manifest: {e}"))
+            .with_context(|| anyhow::anyhow!("{}", manifest_path_buf.display()))?;
+        let mut manifest = wasmer_toml::Manifest::parse(&manifest)?;
+        manifest.base_directory_path = cwd.clone();
+
+        if let Some(package_name) = self.package_name.as_ref() {
+            manifest.package.name = package_name.to_string();
+        }
+
+        if let Some(version) = self.version.as_ref() {
+            manifest.package.version = version.clone();
+        }
+
+        let registry = match self.registry.as_deref() {
+            Some(s) => wasmer_registry::format_graphql(s),
+            None => {
+                let config = PartialWapmConfig::from_file()
+                    .map_err(|e| anyhow::anyhow!("could not load config {e}"))?;
+                config.registry.get_current_registry()
+            }
+        };
+
+        if !self.no_validate {
+            validate::validate_directory(&manifest, &registry, cwd.clone())?;
+        }
+
+        builder.append_path_with_name(&manifest_path_buf, PACKAGE_TOML_FALLBACK_NAME)?;
+
+        let manifest_string = toml::to_string(&manifest)?;
+
+        let (readme, license) = construct_tar_gz(&mut builder, &manifest, &cwd)?;
+
+        builder.finish().ok();
```

Comment:
If finishing the builder fails here, should we be bubbling errors up to the user?

Reply:
Yes, done in https://github.com/wasmerio/wasmer/pull/3432/commits/96f14164af80f1f0b078488f8b6c839508df4d18